Custom import resolution

hjdivad opened the pull request:
- [x] test downstream
- [x] update docs

stefanpenner commented on eslint-plugin/rules/validate-imports.js:
```diff
@@ -16,13 +16,37 @@ module.exports = {
+      context.options.length > 0 &&
+      context.options[0].importResolver;
+    if (typeof importResolver === 'string') {
+      resolveImport = require(importResolver);
```
Dynamic `require` like this has some problems, as it is vulnerable to node_module topology changes, yarn/npm linking etc.  In general, unless we can explicity depend on something, or we can provide the appropriate basedir, we should only allow importing of absolute paths.

hjdivad committed:
Support configurable resolvers

The default import resolution strategy is to use `require.resolve`.
Custom import resolution strategies can be passed in via

  - `--import-resolver` (cli)
  - as `resolveImport` (lib).
  - as the option `importResolver` (eslint)

For the CLI and eslint, the resolver must be a string that can be
`require`d, and whose export is a resolver function.

For the LIB `resolveImport` must be a resolver function.

A resolver function implements

  (
    identifier: string,
    options: { basedir: string, ...args: [any]}
  ) => string

Returning the absolute path to the resolved import, or throwing an error
e with `e.code === MODULE_NOT_FOUND` otherwise.
